cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/SystemMetadataWriterStage.java

```diff
@@ -45,7 +48,10 @@ public void process(ApplicationWithPrograms input) throws Exception {
     // add system metadata for apps
     ApplicationId appId = input.getApplicationId();
     ApplicationSpecification appSpec = input.getSpecification();
-    SystemMetadataWriter appSystemMetadataWriter = new AppSystemMetadataWriter(metadataStore, appId, appSpec);
+    // only update creation time if this is a new app
+    Map<String, String> properties = metadataStore.getProperties(MetadataScope.SYSTEM, appId);
+    SystemMetadataWriter appSystemMetadataWriter =
+      new AppSystemMetadataWriter(metadataStore, appId, appSpec, !properties.isEmpty());
```

Comment:
I see. The hack and potential failure scenario you described can also happen for apps, which is why I didn't think it was an issue.
I suppose for the current approach DatasetAdminService/StreamAdmin have better behavior.